Expose Error function for WrappedError

ozairasim opened the pull request:
<!--
    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --sign-off flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - No tests are needed for internal implementation changes.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

### Description
[WrappedError](https://github.com/prometheus/prometheus/blob/main/pkg/rulefmt/rulefmt.go#L41-L47) is public and is returned from the [RuleNode.Validate()](https://github.com/prometheus/prometheus/blob/main/pkg/rulefmt/rulefmt.go#L134-L135) function, but we do not expose any fields or methods for the WrappedError as a result the callers cannot extract meaningful information from it. This PR exposes an Error() method so the callers can print the error returned

ozairasim commented:
Hi @roidelapluie and @LeviHarrison! I would appreciate your reviews on this PR.

LeviHarrison commented on pkg/rulefmt/rulefmt.go:
```diff
@@ -46,13 +56,14 @@ type WrappedError struct {
-		return errors.Wrapf(err.Err.err, "%d:%d: %d:%d: group %q, rule %d, %q", err.Err.node.Line, err.Err.node.Column, err.Err.nodeAlt.Line, err.Err.nodeAlt.Column, err.Group, err.Rule, err.RuleName).Error()
-	} else if err.Err.node != nil {
-		return errors.Wrapf(err.Err.err, "%d:%d: group %q, rule %d, %q", err.Err.node.Line, err.Err.node.Column, err.Group, err.Rule, err.RuleName).Error()
+// Error prints the error message in formatted string.
```
```suggestion
// Error prints the error message in a formatted string.
```

LeviHarrison commented on pkg/rulefmt/rulefmt.go:
```diff
@@ -38,6 +38,16 @@ type Error struct {
 	Err      WrappedError
 }
 
+// Error prints the error message in formatted string.
```
```suggestion
// Error prints the error message in a formatted string.
```

LeviHarrison commented on pkg/rulefmt/rulefmt_test.go:
```diff
@@ -184,3 +186,120 @@ groups:
+func TestError_Error(t *testing.T) {
+	tests := []struct {
+		name     string
+		Group    string
+		Rule     int
+		RuleName string
+		Err      WrappedError
```
Instead of replicating `Error`'s fields here, I think it would be more concise if we just specified a complete `Error` object.

```suggestion
		Error *Error
```

LeviHarrison commented on pkg/rulefmt/rulefmt_test.go:
```diff
@@ -184,3 +186,120 @@ groups:
+func TestWrappedError_Error(t *testing.T) {
+	tests := []struct {
+		name    string
+		err     error
+		node    *yaml.Node
+		nodeAlt *yaml.Node
```
Same here.

LeviHarrison commented on pkg/rulefmt/rulefmt_test.go:
```diff
@@ -184,3 +186,120 @@ groups:
+	}
+}
+
+func TestWrappedError_Error(t *testing.T) {
```
I don't think we need to prefix `_Error`.

```suggestion
func TestWrappedError(t *testing.T) {
```

LeviHarrison commented on pkg/rulefmt/rulefmt_test.go:
```diff
@@ -184,3 +186,120 @@ groups:
 	require.NotEqual(t, err0, err1, "Error nodes should not be the same")
 }
+
+func TestError_Error(t *testing.T) {
```
```suggestion
func TestError(t *testing.T) {
```

ozairasim commented:
Actually I think we should not need to call Validate on RuleNode, but should rather use the Validate for RuleGroup, hence we might not need this change. However, as the Validate is a public function, I think the PR still makes sense, but I would leave it up to you if you would like to go ahead with merging it :).

I've also addressed your concerns.

LeviHarrison commented:
Just did a quick rebase because `pkg/rulefmt` was moved to `model/rulefmt`.

LeviHarrison reviewed (approved):
I've decided that we should merge this.

Thanks!
